### PR TITLE
Index-Free Queries (disabled, some code + tests missing)

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This PR adds the Index-Free Query Engine.

It also:
- Writes the ReadTime row for all documents added to the RemoteDocumentCache.
- Adds the ability to filter queries by ReadTime to the RemoteDocumentCache. This is not actually invoked yet since Queries don't have their read time set.

### CODE REVIEW

**Comment by schmidt-sebastian** [comment]
@wu-hui I think this is safe to merge. Assigning to Gil for another round of verification.

File: Example/Firebase.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -53,6 +53,7 @@
 /* End PBXAggregateTarget section */
 
 /* Begin PBXBuildFile section */
+		02423E46B05808D7F318DCE7 /* Pods_Storage_Example_iOS.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 270F2E600F4EE5C1A758C5EB /* Pods_Storage_Example_iOS.framework */; };
```
Comment (wilhuff) [inline Example/Firebase.xcodeproj/project.pbxproj]: This doesn't belong.

File: Example/Storage/Tests/Integration/FIRStorageIntegrationTests.m [hunk 2]
```diff
@@ -64,6 +64,7 @@ - (void)setUp {
 
   self.app = [FIRApp defaultApp];
   self.storage = [FIRStorage storageForApp:self.app];
+  self.storage.callbackQueue = nil;
```
Comment (wilhuff) [inline Example/Storage/Tests/Integration/FIRStorageIntegrationTests.m]: This also doesn't belong.

File: Firestore/core/src/firebase/firestore/local/CMakeLists.txt [hunk 3]
```diff
@@ -77,6 +77,14 @@ cc_library(
 cc_library(
   firebase_firestore_local
   SOURCES
+<<<<<<< HEAD
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/local/CMakeLists.txt]: Merge conflict.

File: Firestore/core/src/firebase/firestore/local/remote_document_cache.h [hunk 4]
```diff
@@ -82,9 +84,13 @@ class RemoteDocumentCache {
    * Cached DeletedDocument entries have no bearing on query results.
    *
    * @param query The query to match documents against.
+   * @param since_read_time If not set to SnapshotVersion::NONE(), return only
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/local/remote_document_cache.h]: SnapshotVersion::None(), none NONE.

File: Firestore/core/test/firebase/firestore/local/CMakeLists.txt [hunk 5]
```diff
@@ -28,6 +28,15 @@ cc_test(
     leveldb_transaction_test.cc
     leveldb_util_test.cc
     local_serializer_test.cc
+<<<<<<< HEAD
```
Comment (wilhuff) [inline Firestore/core/test/firebase/firestore/local/CMakeLists.txt]: Merge artifact.

File: Firestore/core/src/firebase/firestore/local/CMakeLists.txt [hunk 6]
```diff
@@ -77,6 +77,11 @@ cc_library(
 cc_library(
   firebase_firestore_local
   SOURCES
+    document_key_reference.cc
+    document_key_reference.h
+    index_free_query_engine.h
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/local/CMakeLists.txt]: Please sort contents (.cc should be before .h)

Comment (schmidt-sebastian) [reply Firestore/core/src/firebase/firestore/local/CMakeLists.txt]: Done

**Pull request merged**